### ISSUE TITLE
False positive on unit_expr involving `!`

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/rust-clippy#2330]
I consider this a false positive. The suggested `process::exit(1)` is not better than `process::exit(1);`.

```rust
use std::env;
use std::process;

fn main() {
    let mut args = env::args();
    let _ = args.next(); // executable name

    let filename = match args.next() {
        Some(filename) => filename,
        None => {
            eprintln!("Usage: false-positive path/to/filename.rs");
            process::exit(1);
        }
    };

    println!("{}", filename);
}
```

```
warning: This expression evaluates to the Unit type ()
  --> src/main.rs:8:20
   |
8  |       let filename = match args.next() {
   |  ____________________^
9  | |         Some(filename) => filename,
10 | |         None => {
11 | |             eprintln!("Usage: false-positive path/to/filename.rs");
12 | |             process::exit(1);
13 | |         }
14 | |     };
   | |_____^
   |
   = note: #[warn(unit_expr)] on by default
note: Consider removing the trailing semicolon
  --> src/main.rs:12:13
   |
12 |             process::exit(1);
   |             ^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.177/index.html#unit_expr
```

**Comment by HMPerson1** [comment]
IMHO, the current `unit_expr` lint (in src/is_unit_expr.rs) should be removed. It's written as an `EarlyLintPass`, so it doesn't use type information at all, and it's already caused other problems (#2303). There are already type-aware lints that cover most cases of doing useless stuff with units (`no_effect`, which detects `"foo";`; and `let_unit_value` which detects `let x = ();`). The only thing that's not already covered is passing a unit value to a function.

If everyone is ok with this, I'll work on a PR that adds a lint for that last one and removes `unit_expr`.

**Comment by oli-obk** [comment]
@HMPerson1 yea that seems sensible.

**Comment by Manishearth** [comment]
Agreed.
